### PR TITLE
Set/Object Comprehensions

### PR DESCRIPTION
**Pull request by mmussomele**
Implement Set and Object comprehensions, as well as a fixes for a few discovered bugs.

### CODE REVIEW

File: topdown/topdown.go [hunk 1]
```diff
@@ -1932,7 +1907,34 @@ func evalTermsComprehension(t *Topdown, comp ast.Value, iter Iterator) error {
...
+		}
 
+		return Continue(t, comp, &result, iter)
+	case *ast.ObjectComprehension:
+		result := ast.Object{}
+		child := t.Closure(comp.Body)
+		err := Eval(child, func(child *Topdown) error {
+			key := PlugTerm(comp.Key, child.Binding)
+			value := PlugTerm(comp.Value, child.Binding)
+			result = append(result, [2]*ast.Term{key, value})
```
Comment (tsandall) [inline topdown/topdown.go]: We should check for conflicts here. For example:

```
s = {["a", 1], ["b", 2], ["a", 3]}
m = {k: v | s[x]; x = [k, v]}
```

Need to update the docs as well to explain this behaviour.

File: topdown/topdown_test.go [hunk 2]
```diff
@@ -940,16 +940,38 @@ func TestTopDownComprehensions(t *testing.T) {
...
+		{"object simple", []string{`p[i] { xs = {s: x | x = a[_]; format_int(x, 10, s)}; y = xs[i]; y > 1 }`}, `["2","3","4"]`},
+		{"object nested", []string{`p = r { r = {x: y | z = {i: q | i = b[q]}; x = z[y]}}`}, `{"v1": "hello", "v2": "goodbye"}`},
+		{"object embedded array", []string{`p[i] { xs = [{s: x | x = a[_]; format_int(x, 10, s)}]; xs[0][i] > 1 }`}, `["2","3","4"]`},
+		{"object embedded object", []string{`p[i] { xs = {"a": {s: x | x = a[_]; format_int(x, 10, s)}}; xs.a[i] > 1 }`}, `["2","3","4"]`},
+		{"object embedded set", []string{`p = xs { xs = {{s: x | x = a[_]; format_int(x, 10, s)}} }`}, `[{"1":1,"2":2,"3":3,"4":4}]`},
+		{"object closure", []string{`p[x] { y = 1; x = {"foo":y | y = 1} }`}, `[{"foo": 1}]`},
+		{"object dereference embedded", []string{
+			`p[x] { q.a = x }`,
+			`q[k] = v { k = "a"; v = {"bar": y | i[_] = _; i = y; i = {"foo": z | z = a[_]}} }`,
+		}, `[{"bar": {"foo": 4}}]`},
```
Comment (tsandall) [inline topdown/topdown_test.go]: Need to test conflicts here.

File: ast/compile.go [hunk 3]
```diff
@@ -1573,6 +1579,46 @@ func (vis *bodySafetyVisitor) checkArrayComprehensionSafety(ac *ArrayComprehensi
 	}
 }
 
+func (vis *bodySafetyVisitor) checkObjectComprehensionSafety(oc *ObjectComprehension) {
```
Comment (tsandall) [inline ast/compile.go]: These three checks share a lot of code. Did you think about refactoring to avoid duplication?

File: ast/env.go [hunk 4]
```diff
@@ -107,14 +107,31 @@ func (env *TypeEnv) Get(x interface{}) types.Type {
...
 		if len(errs) == 0 {
 			return types.NewArray(nil, cpy.Get(x.Term))
 		}
 
 		return nil
+	case *ObjectComprehension:
+		checker := newTypeChecker()
+		cpy, errs := checker.CheckBody(env, x.Body)
+		if len(errs) == 0 {
+			// TODO(mmussomele): How to set an object comprehension's
```
Comment (tsandall) [inline ast/env.go]: I think comment is out-of-date. We need to come back to add key types for both objects and object comprehensions though.

File: topdown/topdown.go [hunk 5]
```diff
@@ -1898,7 +1898,8 @@ func evalTermsComprehension(t *Topdown, comp ast.Value, iter Iterator) error {
 		child := t.Closure(comp.Body)
 
 		err := Eval(child, func(child *Topdown) error {
-			result = append(result, PlugTerm(comp.Term, child.Binding))
+			t := expandTermRef(child, comp.Term)
```
Comment (tsandall) [inline topdown/topdown.go]: I would change this to:

1. Plug the term to replace values w/ bindings (`PlugTerm`)
2. Resolve refs to replace local var refs w/ values (`ResolveRefs`)

This should remove the need for `expandTermRef` here.

File: docs/book/how-do-i-write-policies.md [hunk 6]
```diff
@@ -579,7 +579,61 @@ The result:
 +-----------+------------------------------------------------------+
 ```
 
-In the future, Rego will support Set and Object comprehensions.
+### <a name="object-comprehensions"></a> Object Comprehensions
```
Comment (tsandall) [inline docs/book/how-do-i-write-policies.md]: Need to update SUMMARY.md to include these.

**Comment by mmussomele** [comment]
Ready for another look.

File: topdown/topdown.go [hunk 7]
```diff
@@ -1948,7 +1923,42 @@ func evalTermsComprehension(t *Topdown, comp ast.Value, iter Iterator) error {
 		}
 
 		return Continue(t, comp, result, iter)
+	case *ast.SetComprehension:
+		result := ast.Set{}
+		child := t.Closure(comp.Body)
+
+		err := Eval(child, func(child *Topdown) error {
+			result = append(result, PlugTerm(comp.Term, child.Binding))
```
Comment (tsandall) [inline topdown/topdown.go]: Nit: use ast.Set.Add here instead of append.

File: docs/book/how-do-i-write-policies.md (hunk 6 above)
Comment (tsandall) [inline docs/book/how-do-i-write-policies.md]: Can we include a short paragraph describing/showing conflict behaviour?

**Comment by mmussomele** [comment]
Fixed.

**Comment by tsandall** [comment]
:shipit:

**Pull request merged**